feat(ingestion): run all analytic events through populateTeamDataStep

xvello opened the pull request:
## Problem

Closes #14327, part of removing the Postgres dependency from capture.py

## Changes

- Run all events through the `populateTeamDataStep`, not just events without a team_id. This allows us to:
  - confirm that the Postgres load will be OK once we fully roll it out (and rollback without getting in a pickle)
  - control the rollout from a single point (dropping the team_id field in capture)
- If a `team_id` field is present, trust it but report inconsistencies with the new `ingestion_team_resolution_checks_total` counter
- Emit a new `ingestion_event_dropped_total` counter, that could be used by other parts of the pipeline (dropping events from apps)
- Tune the team cache settings, only cache negative lookups for 5 minutes
- Don't drop the token from the event, mirroring previous behaviour


## How did you test this code?

unit and end2end tests

xvello committed:
team-manager: expire negative lookups after 5 minutes, improve docs

xvello committed:
populateTeamDataStep: don't drop token, keep team_id from capture if present, report results

xvello committed:
ingestEvent: run all analytic events through runLightweightCaptureEndpointEventPipeline

xvello commented on plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts:
```diff
@@ -1,28 +1,69 @@
+    if (event.team_id) {
+        if (team?.id || event.team_id) {
+            inconsistentTeamCounter
+                .labels({
```
High cardinality, but expect to have zero contexts (hopefully?).
We could play it safer and not add the labels for now, only add them if we find it's not zero?

tiina303 replied:
Assuming we expect to see nothing here and if we do, then we can revert this fast so seems fine 🤷‍♀️ 
What's the failure mode: we kill prometheus?

xvello replied:
Indeed, failure mode is we kill prometheus and lose observability and alerting 😅 
Gonna remove the label just to be sure.

xvello commented on plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts:
```diff
@@ -1,28 +1,69 @@
+                    resolved_team_id: team?.id,
+                })
+                .inc()
+            runner.hub.statsd?.increment('ingestion_inconsistent_team_resolution_total')
```
_Ceinture et bretelles_, in case the prometheus metric does not emit due to multi-threading

xvello commented on plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts:
```diff
@@ -33,6 +74,5 @@ export async function populateTeamDataStep(
         ip: team.anonymize_ips ? null : event.ip,
     }
 
-    delete event['token']
```
keeping it as it can be useful through the new kafka partition statistics table.

tiina303 commented on plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts:
```diff
@@ -1,28 +1,69 @@
+     */
+
+    // Events ingested with no token are dropped, they should be blocked by capture
     if (!event.token) {
```
just to sanity check - capture always keeps token regardless if we resolved team-id or not

xvello replied:
Capture does, but import apps probably don't.
Allowing events to come in without a token for now, and adding counters to confirm that.

tiina303 commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -50,26 +49,41 @@ export class TeamManager {
+         * Validates and resolves the api token from an incoming event.
+         *
+         * Caching is added to reduce the load on Postgres, not to be resilient
+         * to failures. If PG is unavailable, this function will trow and the
+         * lookup must be retried later.
```
only if not in cache, right? (clarify the comment)

tiina303 commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -50,26 +49,41 @@ export class TeamManager {
+
         const cachedTeamId = this.tokenToTeamIdCache.get(token)
 
-        // tokenToTeamIdCache.get returns `undefined` if the value doesn't
-        // exist so we check for the value being `null` as that means we've
-        // explictly cached that the team does not exist
```
I liked this comment a bit better than simply `Negative lookups (`null` instead of `undefined`)` - note that we had an problem because we messed this up somewhere else for a while (group properties were not added to events)

xvello committed:
remove labels on inconsistentTeamCounter for first deploy

xvello committed:
update comments

tiina303 reviewed (approved):
lgtm - not sure why the tests failed though, can you check that out first

xvello committed:
update tests

xvello committed:
continue accepting events with no token but a team_id

lots of end2end tests run this way, and this might be how
we capture events from import plugins

tiina303 commented on plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts:
```diff
@@ -1,28 +1,80 @@
+                inconsistentTeamCounter.inc()
+                runner.hub.statsd?.increment('ingestion_inconsistent_team_resolution_total')
+            }
+        }
```
statsd the else case here too

xvello replied:
Changing the prom counter to a `check_ok:true/false` tag, and removing the statsd dupe, as I'm pretty sure we're in the main loop 🤞

xvello committed:
make ingestion report successful team consistency checks too

hazzadous commented on plugin-server/tests/helpers/metrics.ts:
```diff
@@ -0,0 +1,15 @@
+    register.resetMetrics()
+}
+
+export async function getMetricValues(metricName: string): Promise<any> {
```
fwiw we do a similar thing in the functional tests, which would have caught the piscina separate process issue. I need to make these easier to run though :-/

hazzadous commented on plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts:
```diff
@@ -61,22 +60,10 @@ export async function ingestEvent(
-            team_id: event.team_id.toString(),
-        })
-        // we've confirmed team_id exists so can assert event as PluginEvent
-        await workerMethods.runEventPipeline(event as PluginEvent)
```
non-blocking

Is this called anywhere else now? Maybe not for this PR but how easy would it be to remove altogether? Might be a pain with all the test references.

xvello replied:
I tried to, but there's indeed a lot of test references everywhere. I have a branch dropping `runEventPipeline` and renaming `runLightweightCaptureEndpointEventPipeline` to `runEventPipeline`, but that branch has a lot of test breakages.

Want to get that code in prod fast, to confirm out hypothesis, while I cleanup the other branch,

xvello replied:
Draft PR for the cleanups: https://github.com/PostHog/posthog/pull/14365

hazzadous reviewed (approved):
Looks good. Functional tests are passing so pretty confident it's good, I believe there are tests around the affected scenarios.

xvello committed:
mirror metrics in statsd for now
